api/api/src/main/scala/uk/ac/wellcome/platform/api/models/Aggregation.scala

```diff
@@ -69,87 +59,112 @@ object Aggregations extends Logging {
+    }
+
+  implicit val decodeSubjectFromLabel: Decoder[Subject[Minted]] =
+    Decoder.decodeString.map { str =>
+      Subject(label = str, concepts = Nil)
     }
 
-    /** If the `filtered` sub-aggregation is present (see [[uk.ac.wellcome.platform.api.services.FiltersAndAggregationsBuilder]])
-      * then use the count from it */
-    private def getFilteredBucket[T](
-      esAggBucket: EsAggregationBucket[T]): AggregationBucket[T] =
-      AggregationBucket(
-        esAggBucket.key,
-        esAggBucket.filtered
-          .map(_.doc_count)
-          .getOrElse(esAggBucket.doc_count))
-
-    implicit def fromJsonAggReader[T: Decoder]: AggReader[T] =
-      (json: String) => fromJson[T](json)
+  implicit class EnhancedEsAggregations(
+    aggregations: com.sksamuel.elastic4s.requests.searches.Aggregations) {
+
+    def decodeAgg[T: Decoder](
+      name: String,
+      documentPath: Option[String] = None): Option[Aggregation[T]] = {
+      aggregations
+        .getAgg(name)
+        .flatMap(
+          _.safeTo[Aggregation[T]]((json: String) =>
+            AggregationMapping
+              .aggregationParser[T](json, documentPath)).recoverWith {
+            case err =>
+              warn("Failed to parse aggregation from ES", err)
+              Failure(err)
+          }.toOption
+        )
+    }
+  }
+}
+
+// This object contains the tools required to map an aggregation result from ES
+// into our Aggregation data type.
+//
+// There 2 independent variables in each bucket of these aggregations:
+//
+// ## Root count vs filtered count
+// If aggregation buckets have a subaggregation named `filtered` from [[uk.ac.wellcome.platform.api.services.FiltersAndAggregationsBuilder]]
+// then we use the count from there, otherwise we use the count from the root of the bucket.
+//
+// ## Key-only vs sample document
+// For some aggregation types `T` (in `Aggregation[T]`) we can decode `T` directly from the bucket key -
+// for example with WorkType, we can use `fromCode` to get the label of the WorkType given its id.
+// However, sometimes we need more information from the index, which is achieved using a top hits
+// aggregation of size 1 - see [[uk.ac.wellcome.platform.api.services.ElasticsearchQueryBuilder]].
+//
+// In this case there is a `sample_doc` subaggregation for which we can provide a
+// `path` to the relevant data inside it. For example, the Language type uses this to obtain both the
+// label and id at the path `data.language` within `sample_doc`.
+object AggregationMapping extends Logging {
+  import io.circe.parser._
+  import io.circe.optics.JsonPath
+  import io.circe.optics.JsonPath._
+
+  private val buckets = root.buckets.arr
+  private val bucketFilteredCount = root.filtered.doc_count.int
+  private val bucketRootCount = root.doc_count.int
+  private val bucketKey = root.key.json
+  private val bucketSampleDoc =
+    root.sample_doc.hits.hits.index(0)._source.json
+
+  def aggregationParser[T: Decoder](
```

Comment:
I think this method could be split up into a few smaller sub methods to help with readability. 

Also, have just done a grep and surprisingly this is the only place in the whole repo where there are any explicit `return` statements (this might be another indicator that the method is too long). I tend to prefer a coding style in functional languages where everything is just a simple expression which evaluates to a value, but up to you if you think it helps with readability.